ivadomed/losses.py

```diff
@@ -252,3 +254,101 @@ def forward(self, input, target):
         return focal_tversky_sum / n_classes
+
+
+class L2loss(nn.Module):
+    """
+    L2 loss between two images : inputs and target
+    """
+
+    def __init__(self):
+        super(L2loss, self).__init__()
+
+    def forward(self, input, target):
+        return torch.sum((input - target) ** 2) / 2
+
+
+class AdapWingLoss(nn.Module):
+    """
+    adaptive Wing loss : https://arxiv.org/abs/1904.07399
+    Used for heatmap ground truth.
+
+    """
+
+    def __init__(self):
+        super(AdapWingLoss, self).__init__()
+
+    def forward(self,input,target):
+        theta = 0.5
+        alpha = 2.1
+        w = 14
+        e = 1
+        A = w * (1 / (1 + torch.pow(theta / e, alpha - target))) * (alpha - target) * torch.pow(theta / e,
+                                                                                              alpha - target - 1) * (1 / e)
+        C = (theta * A - w * torch.log(1 + torch.pow(theta / e, alpha - target)))
+
+        batch_size = target.size()[0]
+        hm_num = target.size()[1]
+
+        mask = torch.zeros_like(target)
+        # W=10
+        kernel = scipy.ndimage.morphology.generate_binary_structure(2, 2)
+        for i in range(batch_size):
+            img_list = list
+            img_list.append(np.round(target[i].cpu().numpy() * 255))
+            img_merge = np.concatenate(img_list)
+            img_dilate = scipy.ndimage.morphology.binary_opening(img_merge, np.expand_dims(kernel, axis=0))
+            img_dilate[img_dilate < 51] = 1  # 0*W+1
+            img_dilate[img_dilate >= 51] = 11  # 1*W+1
+            img_dilate = np.array(img_dilate, dtype=np.int)
+
+            mask[i] = torch.tensor(img_dilate)
+
+        diff_hm = torch.abs(target - input)
+        AWingLoss = A * diff_hm - C
+        idx = diff_hm < theta
+        AWingLoss[idx] = w * torch.log(1 + torch.pow(diff_hm / e, alpha - target))[idx]
+
+        AWingLoss *= mask
+        sum_loss = torch.sum(AWingLoss)
+        all_pixel = torch.sum(mask)
+        mean_loss = sum_loss  # / all_pixel
+
+        return mean_loss
+
+
+class Loss_Combination(nn.Module):
+    """
+    Loss that sums different other implemented loss
+    """
+
+    def __init__(self, losses_list,params_list=None):
+        """
+
+        Args:
+            losses_list(list): list of losses that will be summed. Elements should be string.
+            params_list(list): list of params for the losses, contain None or dictionnary definition of params for the loss
+            at same index. If no params list is given all default parameter will be used.
+            (e.g., losses_list = ["L2loss","DiceLoss"]
+                   params_list = [None,{"param1:0.5"}])
+        returns:
+            tensor: sum of losses computed on (input,target) with the params
+        """
+        super(Loss_Combination, self).__init__()
+        self.losses_list = losses_list
+        self.params_list = params_list
+
+    def forward(self, input, target):
```

Comment:
add docstrings

Reply:
@lrouhier did you see my comment?

Reply:
Yes sorry I forgot to answer. I can add it if you want, however, there is an automated doc for forward in all losses [here](https://ivadomed.org/en/latest/api_ref.html#module-ivadomed.losses) (there is no docstring for the other forward methods in the doc)

Reply:
cool! how is this auto doc generated?

Reply:
Comes from sphinx According to @Drulex 
https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html

Reply:
yes, i know we are using sphinx, but my question was more: where is the native doc for "forward()" defined, and how is it populated for all the instances of this function&

Reply:
I don't know, it is not defined in our code. There is a similar doc for each model forward method. My best guess is sphinx doing some auto-completion. I am looking to it, and I asked @Drulex if he knows how it is created.

Reply:
This one is a mystery.. Looking..

Reply:
It's from the `PyTorch` documentation. Multiple classes in `losses.py` inherit ` torch.nn.modules.module` which already has a `forward` method. The overridden `forward` method in `losses.py` doesn't have a docstring, but it looks like `Sphinx` uses the docstring from the base class. See https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.forward

Reply:
Thanks for tracking this !

Reply:
awesome! thank you for revealing the mystery @Drulex 
good to know for the future